### PR TITLE
[Cast] Refactor kernel functions

### LINKED ISSUE #441: Kernel refactor for `Cast`

### PR DESCRIPTION
**Pull request by virena**
**Description of changes:**

This PR removes kernel functions from `cast.h` and `cast.cc` into separate `cast_kernels.h`, `cast_kernels.cu`, and `cast_kernels.cpp`.

**Related Issues:**

Linked Issues:
- Issue #303 

Issues closed by this PR:
- Closes #441

**Before merging:**

- [x] Did you update the [flexflow-third-party](https://github.com/flexflow/flexflow-third-party) repo, if modifying any of the Cmake files, the build configs, or the submodules? **n/a**

### CODE REVIEW

File: include/flexflow/ops/cast.h [hunk 1]
```diff
@@ -13,22 +13,13 @@
  * limitations under the License.
  */
 #pragma once
```
Comment (lockshaw) [inline include/flexflow/ops/cast.h:15]: Change to ifdef for consistency (yeah I know you didn't make this change, but just throw it in this PR :hearts:)

File: src/ops/kernels/cast_kernels.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+#include <hip/hip_runtime.h>
+
+namespace FlexFlow {
+
+CastMeta::CastMeta(FFHandler handle) : OpMeta(handle) {}
+
+namespace Kernels {
+namespace Cast {
+
+/*static*/
```
Comment (lockshaw) [inline src/ops/kernels/cast_kernels.cpp:27]:
```suggestion
```

File: src/ops/kernels/cast_kernels.cu [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+/*static*/
+template <typename IDT, typename ODT>
+void forward_kernel_wrapper(const IDT *input_ptr,
+                            ODT *output_ptr,
+                            size_t volume) {
+  cudaStream_t stream;
+  checkCUDA(get_legion_stream(&stream));
+
+  cudaEvent_t t_start, t_end;
+  if (false) {
```
Comment (lockshaw) [inline src/ops/kernels/cast_kernels.cu:35]: Can you create an issue to deal with this when you get a chance?

**Review by lockshaw — CHANGES_REQUESTED** [review]
Shouldn't fused changes also be necessary?

**Comment by virena** [comment]
> Shouldn't fused changes also be necessary?

`OP_CAST` is not originally in fused, why would we need it?

File: src/ops/kernels/cast_kernels.cu (hunk 3 above)
Comment (virena) [reply src/ops/kernels/cast_kernels.cu:35]: #529

This is what you were talking about right? I feel comfortable getting rid of these blocks in this PR.

Comment (lockshaw) [reply src/ops/kernels/cast_kernels.cu]: It should probably be converted to https://github.com/flexflow/FlexFlow/blob/f3f441c3ad731b45204b5f76f3d7aecee85e0288/src/ops/kernels/linear_kernels.cu#L94-L98

**Review by lockshaw — APPROVED** [review]

**Comment by lockshaw** [comment]
> > Shouldn't fused changes also be necessary?
> 
> `OP_CAST` is not originally in fused, why would we need it?

Tracking in #571

**Comment by jiazhihao** [comment]
We should include cast in fused_op, but this can be done in a separate PR.

**Pull request merged**